crates/uv-resolver/src/resolver/mod.rs

```diff
@@ -1854,17 +1869,30 @@ enum Response {
     },
 }
 
-/// An enum used by [`DependencyProvider`] that holds information about package dependencies.
-/// For each [Package] there is a set of versions allowed as a dependency.
+/// Information about the dependencies for a particular package.
+///
+/// This effectively distills the dependency metadata of a package down into
+/// its pubgrub specific constituent parts: each dependency package has a range
+/// of possible versions.
 #[derive(Clone)]
 enum Dependencies {
     /// Package dependencies are not available.
     Unavailable(UnavailableVersion),
     /// Container for all available package versions.
+    ///
+    /// Note that in universal mode, it is possible and allowed for multiple
+    /// `PubGrubPackage` values in this list to have the same package name.
+    /// These conflicts are resolved via `Dependencies::fork`.
```

Comment:
Can multiple dependencies of the same name also exist without forking, e.g. when foo has:

```
Requires-Dist: bar >= 1
Requires-Dist: bar < 2
```

These cases exist e.g. with extras, but sometimes also with non-conflicting environment markers.

Reply:
Yeah great question. I was careful to make `Dependencies::fork` infallible so that it didn't try to be too eager about reporting errors. The idea is that if there aren't any requirements with the same package with non-overlapping markers, then the result returned by `Dependencies::fork` is unchanged from what it was. It's just a flat list of requirements. That is, no forks. So something like the scenario you outlined will work just as it always has.

One hiccup here though is if you have this:

```
Requires-Dist: bar >= 1 ; sys_platform == 'linux'
Requires-Dist: bar < 2 ; sys_platform == 'darwin'
```

Even though the version constraints are non-conflicting, because the marker expressions are non-overlapping, this will actually provoke a fork. The fork means that the `bar >= 1` requirement is free to choose, say, `bar==2.0.0` if it's available because the fork exists in isolation from the `bar < 2` requirement. However, if no fork were to happen, then both the `bar >= 1` and `bar < 2` requirements would be active simultaneously such that `bar==2.0.0` couldn't be chosen. So in this case, there exists a resolution without forking, but it is distinct from the resolution we get if we fork.

I think we could detect this case by looking at whether the version constraints themselves are conflicting. That is, if there exists an overlap in the version constraints, then we don't fork even if the marker expressions are non-overlapping. I think a decision like that would be predicated on the idea that it could be _possible_ to find one resolution that works across multiple platforms. But it might not be possible. It feels like forking in this case, even if it isn't always necessary, results in more freedom for the resolver to find a resolution. But... maybe it's surprising to users in real cases. I'm not sure.

I added tests for some of these cases here: https://github.com/astral-sh/packse/pull/196

Reply:
I created an issue for the non-conflicting requirements with non-overlapping markers scenario: https://github.com/astral-sh/uv/issues/4357